src/textual/command.py

```diff
@@ -105,7 +111,57 @@ def __post_init__(self) -> None:
                 )
 
 
-Hits: TypeAlias = AsyncIterator[Hit]
+@dataclass
+class DiscoveryHit:
+    """Holds the details of a single command search hit."""
+
+    display: RenderableType
+    """A string or Rich renderable representation of the hit."""
+
+    command: IgnoreReturnCallbackType
+    """The function to call when the command is chosen."""
+
+    text: str | None = None
+    """The command text associated with the hit, as plain text.
+
+    If `display` is not simple text, this attribute should be provided by
+    the [Provider][textual.command.Provider] object.
+    """
+
+    help: str | None = None
+    """Optional help text for the command."""
+
+    @property
+    def prompt(self) -> RenderableType:
+        """The prompt to use when displaying the discovery hit in the command palette."""
+        return self.display
+
+    def __lt__(self, other: object) -> bool:
+        if isinstance(other, DiscoveryHit):
+            assert self.text is not None
+            assert other.text is not None
+            return other.text < self.text
+        return NotImplemented
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, Hit):
+            return self.text == other.text
+        return NotImplemented
+
+    def __post_init__(self) -> None:
+        """Ensure 'text' is populated."""
+        if self.text is None:
+            if isinstance(self.display, str):
+                self.text = self.display
+            elif isinstance(self.display, Text):
+                self.text = self.display.plain
+            else:
+                raise ValueError(
+                    "A value for 'text' is required if 'display' is not a str or Text"
+                )
+
+
+Hits: TypeAlias = AsyncIterator["DiscoveryHit | Hit"]
```

Comment:
Shouldn't the union be outside?
Or can it happen that we're yielding both regular hits and discovery hits?

```suggestion
Hits: TypeAlias = AsyncIterator["DiscoveryHit"] | AsyncIterator["Hit"]
```

Reply:
Actually that's probably more accurate (although there was a moment while I was working on this where the version I have above was more accurate; but it's settled in a place where this makes more sense now).

Reply:
Tinkering with that for a moment... doesn't actually work and has a lot of knock-on implications for the typing of the search machinery in the command palette. I think to make this work nicely there'd need to be a fully-parallel discovery pipeline, whereas at the moment I'm making full use of the pre-existing search code. Either that or to make the above work there'd need to be a lot of overloading done, perhaps.

Technically what you say here *is* correct, we should never be building up a mixture of the two.

Reply:
Ok, understood. If being marginally “more correct” increases the burden on the programmer disproportionally, I won't push for this change.